cmd/kubectl-gadget/seccompadvisor.go

```diff
@@ -0,0 +1,154 @@
+	SilenceUsage: true,
+}
+
+var seccompAdvisorStopCmd = &cobra.Command{
+	Use:   "stop",
+	Short: "Stop monitoring and report the policies",
+	RunE:   runSeccompAdvisorStop,
+	SilenceUsage: true,
+}
+
+var seccompAdvisorListCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List existing seccomp traces",
+	RunE:   runSeccompAdvisorList,
+	SilenceUsage: true,
+}
+
+var (
+	outputMode string
+	seccompProfileName string
+)
+
+func init() {
+	// Add generic information.
+	rootCmd.AddCommand(seccompAdvisorCmd)
+	utils.AddCommonFlags(seccompAdvisorCmd, &params)
+
+	seccompAdvisorCmd.AddCommand(seccompAdvisorStartCmd)
+	seccompAdvisorStartCmd.PersistentFlags().StringVarP(&outputMode, "output-mode", "m", "Status", "The trace output mode, possibles values are ExternalResource or Status (default).")
+	seccompAdvisorStartCmd.PersistentFlags().StringVar(&seccompProfileName, "seccomp-profile-name", "", "The seccomp profile which will be created when using ExternalResource as value for --output-mode.")
+
+	seccompAdvisorCmd.AddCommand(seccompAdvisorStopCmd)
+	seccompAdvisorCmd.AddCommand(seccompAdvisorListCmd)
+}
+
+// runSeccompAdvisorStart starts monitoring of syscalls for the given
+// parameters.
+func runSeccompAdvisorStart(cmd *cobra.Command, args []string) error {
+	if params.Podname == "" {
+		return errors.New("Usage: kubectl gadget seccompadvisor start -p podname")
+	}
+
+	config := &utils.TraceConfig{
+		GadgetName:      "seccomp",
+		Operation:       "start",
+		TraceOutputMode: outputMode,
+		TraceOutput:     seccompProfileName,
+		CommonFlags:     &params,
+	}
+
+	traceID, err := utils.CreateTrace(config)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("%s\n", traceID)
+
+	return nil
+}
+
+// runSeccompAdvisorStop reports an already running trace which ID was given
+// as parameter.
+func runSeccompAdvisorStop(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		return errors.New("Usage: kubectl gadget seccomp-advisor stop global-trace-id\n")
+	}
+
+	callback := func(results []gadgetv1alpha1.Trace) error {
+		for _, i := range results {
+			if i.Spec.OutputMode == "ExternalResource" {
+				fmt.Printf("Succesfully created seccomp profile %q!\n", i.Spec.Output)
+			} else {
+				fmt.Printf("%v\n", i.Status.Output)
+			}
+		}
+
+		return nil
+	}
+
+	traceID := args[0]
+
+	err := utils.SetTraceOperation(traceID, "generate")
+	if err != nil {
+		return err
+	}
+
+	// We stop the trace so its Status.State become Stopped.
+	// Indeed, generate operation does not change value of Status.State.
+	err = utils.SetTraceOperation(traceID, "stop")
+	if err != nil {
+		return err
+	}
+
+	err = utils.PrintTraceOutputFromStatus(traceID, "Stopped", callback)
+	if err != nil {
+		return err
+	}
+
+	// WARNING We should defer this call, but trace can sadly timeout, so it is
+	// better for the user to call this function again to get expected result and
+	// delete the trace at the same time.
+	return utils.DeleteTrace(traceID)
```

Comment:
I still found paths where it is impossible to delete the trace. For instance:

1. Using non-existing namespace/pod/container.
2. Using --output-mode=ExternalResource and pod terminate before stopping the gadget.
3. Using --output-mode=ExternalResource and the user does not specify the seccomp-profile-name.

Reply:
To solve this problem, I should do a `defer utils.DeleteTrace(traceID)` (and this was done in a previous version of the series).
The problem in this case is the risk of having the trace which timeout and so user will loose the trace before getting any output and will not be able to run the trace again to get output.
So, I decided to remove the `defer` because I thought it was better for user to run again the trace to get what he/she wants at the cost of deleting the trace him/herself using `kubectl delete trace -n gadget trace-name`.

I agree the good solution would be to deal with the root cause of having the trace timeout, this is why I am currently modifying `waitForOutput()` to make it, I think, better.

For the moment, I will try to add a check on returned error and not delete the trace if this is a timeout error.

Reply:
> I agree the good solution would be to deal with the root cause of having the trace timeout, this is why I am currently modifying `waitForOutput()` to make it, I think, better.
>
Mmm but even if you modify `waitForOutput()` to use `Watch()` instead of polling, you will still have to set a maximum time to wait, isn't it? If so, the probability of having a timeout will be still there. Otherwise, what am I missing?

I tried to reproduce the timeout issue and I couldn't after several tests in different testing clusters. As you said, it is important also to understand the root cause so that we can decide how to proceed. If we confirm it is completely outside our scope, I will definitely just "let it happen". We cannot cover all the cases.

IMO, I would prefer that the user has to re-execute the gadget because the `stop` operation lasted longer than expected (which would be probably an issue outside our scope), instead of introducing a potential issue that the user cannot solve using the CLI itself but executing manually `kubectl delete trace -n gadget trace-name`. Considering also that a normal user does not even know that the CLI is deploying and dealing with traces.

What does other folks think?

Reply:
I think leaking a trace is not a possibility. If the user created a trace using the CLI then they should be able to remove it by using the CLI too.